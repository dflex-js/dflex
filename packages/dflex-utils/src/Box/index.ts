export { default as AbstractBox } from "./AbstractBox";
export { default as Box } from "./Box";
export { default as BoxBool } from "./BoxBool";
export { default as BoxNum } from "./BoxNum";
export { default as BoxRect } from "./BoxRect";
export type { AbstractBoxRect as BoxRectAbstract } from "./BoxRect";
