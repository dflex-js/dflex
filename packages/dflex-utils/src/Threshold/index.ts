export { default as DFlexThreshold } from "./Threshold";

export type { ThresholdPercentages } from "./Threshold";
