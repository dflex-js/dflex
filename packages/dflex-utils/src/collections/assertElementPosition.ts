import type { BoxRectAbstract } from "../Box";

let didThrowError = false;

function assertElementPosition(DOM: HTMLElement, rect: BoxRectAbstract): void {
  if (didThrowError) {
    return;
  }

  const DOMRect = DOM.getBoundingClientRect();

  const keys = Object.keys(rect) as (keyof typeof rect)[];

  keys.forEach((k) => {
    if (
      Object.prototype.hasOwnProperty.call(DOMRect, k) &&
      DOMRect[k] !== rect[k]
    ) {
      didThrowError = true;

      throw new Error(
        `Element position assertion failed. Expected: ${DOMRect[k]} found: ${rect[k]}`
      );
    }
  });
}

export default assertElementPosition;
