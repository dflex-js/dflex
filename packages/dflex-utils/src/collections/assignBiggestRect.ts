import type { RectBoundaries } from "../types";

/**
 * Mutate the rect boundaries object to biggest rectangle if it is bigger than
 * the current one.
 */
function dirtyAssignBiggestRect($: RectBoundaries, elm: RectBoundaries) {
  const { top, left, right, bottom } = elm;

  if (left < $.left) {
    $.left = left;
  }

  if (top < $.top) {
    $.top = top;
  }

  if (right > $.right) {
    $.right = right;
  }

  if (bottom > $.bottom) {
    $.bottom = bottom;
  }
}

export default dirtyAssignBiggestRect;
