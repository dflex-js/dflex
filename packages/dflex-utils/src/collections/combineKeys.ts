type KY = number | string;

function combineKeys(k1: KY, k2: KY) {
  return `${k1}_${k2}`;
}

export default combineKeys;
