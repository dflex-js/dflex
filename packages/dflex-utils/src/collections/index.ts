export { default as combineKeys } from "./combineKeys";
export { default as warnOnce } from "./warnOnce";
export { default as assertElementPosition } from "./assertElementPosition";
export {
  getElmComputedStyle,
  clearComputedStyleMap,
  setRelativePosition,
  setFixedWidth,
  getElmComputedDimensions,
} from "./computedStyleUtils";
