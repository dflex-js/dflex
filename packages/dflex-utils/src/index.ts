export { AxesPoint, Point, PointNum, PointBool } from "./Point";
export { AbstractBox, Box, BoxBool, BoxNum, BoxRect } from "./Box";
export type { AbstractBoxRect } from "./Box";

export { DFlexThreshold as Threshold } from "./Threshold";
export type { ThresholdPercentages } from "./Threshold";

export { Tracker, PREFIX_CYCLE, PREFIX_ID, PREFIX_KY } from "./Tracker";

export { default as TaskQueue } from "./TaskQueue";
export { default as eventDebounce } from "./eventDebounce";

export { DFlexCycle } from "./DFlexCycle";
export type { AbstractDFlexCycle } from "./DFlexCycle";

export type {
  Dimensions,
  Axis,
  Axes,
  Direction,
  AnimationOpts,
  CubicBezier,
  CSSStyle,
  CSSClass,
  CSS,
} from "./types";
export { BOTH_AXIS } from "./types";

export {
  noop,
  combineKeys,
  warnOnce,
  assertElmPos,
  getDimensionTypeByAxis,
  getStartingPointByAxis,
  getEndingPointByAxis,
  getOppositeAxis,
  getAnimationOptions,
} from "./collections";

export {
  getCachedComputedStyleProperty,
  clearComputedStyleCache,
  setStyleProperty,
  removeStyleProperty,
  setRelativePosition,
  setFixedDimensions,
  getElmComputedDimensions,
  getElmPos,
  getElmOverflow,
  getParsedElmTransform,
  hasCSSTransition,
  rmEmptyAttr,
} from "./computedStyleUtils";

export type { RAFFunction, RAFCleanup } from "./environment";

export {
  canUseDOM,
  getSelection,
  getParentElm,
  updateElmDatasetGrid,
  DFlexCreateRAF as createRAF,
  DFlexCreateTimeout as createTimeout,
} from "./environment";

export * as featureFlags from "./FeatureFlags";
