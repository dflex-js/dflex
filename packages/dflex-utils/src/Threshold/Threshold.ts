/* eslint-disable max-classes-per-file */
import { PointNum } from "../Point";

import type { Dimensions, Axis, Direction } from "../types";

import { AbstractBox, BoxBool, BoxNum, AbstractBoxRect } from "../Box";

export interface ThresholdPercentages {
  /** vertical threshold in percentage from 0-100 */
  vertical: number;

  /** horizontal threshold in percentage from 0-100 */
  horizontal: number;
}

class DFlexThreshold {
  readonly thresholds: Record<string, BoxNum>;

  private _pixels!: PointNum;

  private _percentages: ThresholdPercentages;

  isOut: Record<string, BoxBool>;

  constructor(percentages: ThresholdPercentages) {
    this._percentages = percentages;
    this.thresholds = {};
    this.isOut = {};
  }

  private _createPixels({ width, height }: Dimensions): void {
    const x = Math.round((this._percentages.horizontal * width) / 100);
    const y = Math.round((this._percentages.vertical * height) / 100);

    this._pixels = new PointNum(x, y);
  }

  /** Assign threshold property and create new instance for is out indicators */
  private _createThreshold(
    key: string,
    box: AbstractBox,
    isInner: boolean
  ): void {
    if (__DEV__) {
      if (this.thresholds[key] || this.isOut[key]) {
        throw new Error(`Threshold with key: ${key} already exists`);
      }
    }

    this.thresholds[key] = this._pixels.composeBox(box, isInner);
    console.log("ðŸš€ ~ file: Threshold.ts:51 ~ DFlexThreshold ~ box:", box);

    this.isOut[key] = new BoxBool(false, false, false, false);
  }

  private _setDepthThreshold(key: string, depth: number): void {
    const dp = `${depth}`;

    if (!this.thresholds[dp]) {
      this._createThreshold(dp, this.thresholds[key], false);

      return;
    }

    this.thresholds[depth].assignBiggestBox(this.thresholds[key]);
  }

  /**
   * Set the main threshold for the element based on the element's dimensions
   * and threshold types. For dragged and containers threshold type is outer
   * `isInner=false` and for the rest of the elements `isInner=true.`
   * Note: Duplicate threshold keys will throw an error.
   *
   * @param key
   * @param box
   * @param isInner
   */
  setMainThreshold(key: string, box: AbstractBoxRect, isInner: boolean): void {
    this._createPixels(box);

    this._createThreshold(key, box, isInner);
  }

  /**
   * Update existing threshold with new dimensions.
   *
   * @param key
   * @param rect
   * @param isInner
   */
  updateMainThreshold(key: string, rect: AbstractBox, isInner: boolean): void {
    if (__DEV__) {
      if (!this.thresholds[key]) {
        throw new Error(`Threshold ${key} does not exist.`);
      }
    }

    this.thresholds[key] = this._pixels.composeBox(rect, isInner);

    this.isOut[key].setFalsy();
  }

  getElmMainThreshold(rect: AbstractBox): BoxNum {
    return this._pixels.composeBox(rect, false);
  }

  /**
   * Assign outer threshold for the container. Along with another threshold
   * called insertion threshold which defines the area where the element can be
   * inserted during the migration taking into consideration the biggest hight
   * and width for the depth by using `unifiedContainerDimensions`. And create
   * accumulated depth threshold.
   *
   * @param SK
   * @param childDepth
   * @param containerRect
   * @param unifiedContainerDimensions
   */
  setContainerThreshold(
    SK: string,
    insertionLayerKey: string,
    childDepth: number,
    containerRect: AbstractBox,
    unifiedContainerDimensions: Dimensions
  ): void {
    // Regular threshold.
    this._createThreshold(SK, containerRect, false);

    const { top, left } = containerRect;
    const { height, width } = unifiedContainerDimensions;

    // Insertion threshold.
    this._createThreshold(
      insertionLayerKey,
      {
        left,
        top,
        right: left + width,
        bottom: top + height,
      },
      false
    );

    // Accumulated depth threshold. Accumulation based on insertion layer.
    this._setDepthThreshold(insertionLayerKey, childDepth);
  }

  /**
   * Check if the element is out of the threshold from given axis.
   *
   * Note: This method checks and sets so the result is always preserved.
   *
   *
   * @param axis
   * @param key
   * @param startingPos
   * @param endingPos
   * @returns
   */
  isOutThresholdByAxis(
    axis: Axis,
    key: string,
    startingPos: number,
    endingPos: number
  ): boolean {
    const { left, right, top, bottom } = this.thresholds[key];

    if (axis === "x") {
      this.isOut[key].setByAxis(axis, startingPos < left, endingPos > right);
    }

    if (axis === "y") {
      this.isOut[key].setByAxis(axis, startingPos < top, endingPos > bottom);
    }

    return this.isOut[key].isOneTruthyByAxis(axis);
  }

  /**
   * Check if the element is out of the threshold from one direction based only
   * on axis and element direction.
   *
   * Note: This method checks and sets so the result is always preserved
   *
   * @param axis
   * @param direction
   * @param key
   * @param startingPos
   * @param endingPos
   * @returns
   */
  isOutThresholdByDirection(
    axis: Axis,
    direction: Direction,
    key: string,
    startingPos: number,
    endingPos: number
  ): boolean {
    const { left, right, top, bottom } = this.thresholds[key];

    const is =
      axis === "x"
        ? direction === -1
          ? startingPos < left
          : endingPos > right
        : direction === -1
        ? startingPos < top
        : endingPos > bottom;

    this.isOut[key].setOne(axis, direction, is);

    return is;
  }

  isOutThreshold(key: string, box: BoxNum, isPreserveResult: boolean): boolean {
    const thresholdBox = this.thresholds[key];

    return box.isOutThreshold(
      thresholdBox,
      isPreserveResult ? this.isOut[key] : undefined
    );
  }

  destroy(): void {
    Object.keys(this.thresholds).forEach((key) => {
      delete this.thresholds[key];
    });
    Object.keys(this.isOut).forEach((key) => {
      delete this.isOut[key];
    });
  }
}

export default DFlexThreshold;
