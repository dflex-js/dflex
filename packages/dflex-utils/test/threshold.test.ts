import { combineKeys } from "../src/collections";
import { DFlexThreshold } from "../src/Threshold";
import type { Dimensions } from "../src/types";

describe("DFlexThreshold", () => {
  const DEFAULT_IS_OUT_INDICATORS = {
    top: false,
    right: false,
    bottom: false,
    left: false,
  };

  describe("Dragged Threshold", () => {
    let threshold: DFlexThreshold;

    const DRAGGED = {
      ID: "dflex-dragged",
      DP: 0,
      RECT: {
        height: 50,
        width: 170,
        top: 114,
        right: 620,
        bottom: 164,
        left: 450,
      },
      THRESHOLD: {
        top: 84,
        bottom: 194,
        left: 348,
        right: 722,
      },
    };

    const CONTAINER = {
      SK: "container-sk",
      RECT: {
        height: 50,
        width: 170,
        top: 114,
        right: 192,
        bottom: 164,
        left: 22,
      },
      THRESHOLD: {
        bottom: 194,
        left: -80,
        right: 294,
        top: 84,
      },
    };

    const DEPTH = {
      THRESHOLD: {
        // heigh: 518,
        // width: 643,
        top: 54,
        right: 450,
        bottom: 572,
        left: -182,
      },
    };

    const INSERTION_LAYER = {
      key: combineKeys(DRAGGED.DP, CONTAINER.SK),
      THRESHOLD: { bottom: 542, left: -80, right: 348, top: 84 },
    };

    const UNIFIED_CONTAINER_DIMENSION: Dimensions = {
      height: 398,
      width: 224,
    };

    const pixels = {
      x: 102,
      y: 30,
    };

    beforeAll(() => {
      const thresholdInput = {
        vertical: 60,
        horizontal: 60,
      };

      threshold = new DFlexThreshold(thresholdInput);
    });

    afterAll(() => {
      threshold.destroy();
    });

    describe("Main threshold", () => {
      it("Should be defined before setting the keys", () => {
        expect(threshold).toBeDefined();
        expect(threshold.isOut).toStrictEqual({});
        expect(threshold.thresholds).toStrictEqual({});
      });

      it("Add main outer threshold for dragged", () => {
        threshold.setMainThreshold(DRAGGED.ID, DRAGGED.RECT, false);

        expect(threshold.isOut[DRAGGED.ID]).toMatchObject(
          DEFAULT_IS_OUT_INDICATORS
        );

        expect(threshold.thresholds[DRAGGED.ID]).toStrictEqual(
          DRAGGED.THRESHOLD
        );
      });

      it("Outer threshold is bigger than input rect", () => {
        expect(threshold.thresholds[DRAGGED.ID].bottom).toBeGreaterThan(
          DRAGGED.RECT.top + DRAGGED.RECT.height
        );

        expect(threshold.thresholds[DRAGGED.ID].top).toBeLessThan(
          DRAGGED.RECT.top
        );

        expect(threshold.thresholds[DRAGGED.ID].left).toBeLessThan(
          DRAGGED.RECT.left
        );

        expect(threshold.thresholds[DRAGGED.ID].right).toBeGreaterThan(
          DRAGGED.RECT.left + DRAGGED.RECT.width
        );
      });

      it("Calculate the threshold correctly", () => {
        expect(DRAGGED.RECT.top + DRAGGED.RECT.height + pixels.y).toBe(
          threshold.thresholds[DRAGGED.ID].bottom
        );

        expect(DRAGGED.RECT.top - pixels.y).toBe(
          threshold.thresholds[DRAGGED.ID].top
        );

        expect(DRAGGED.RECT.left + DRAGGED.RECT.width + pixels.x).toBe(
          threshold.thresholds[DRAGGED.ID].right
        );

        expect(DRAGGED.RECT.left - pixels.x).toBe(
          threshold.thresholds[DRAGGED.ID].left
        );
      });

      it("Throw when adding the same key", () => {
        expect(() => {
          threshold.setMainThreshold(DRAGGED.ID, DRAGGED.RECT, false);
        }).toThrow();
      });
    });

    describe("Container threshold", () => {
      it("Add threshold for container, it's inner by default", () => {
        threshold.setContainerThreshold(
          CONTAINER.SK,
          INSERTION_LAYER.key,
          DRAGGED.DP,
          CONTAINER.RECT,
          UNIFIED_CONTAINER_DIMENSION
        );

        expect(threshold).toMatchSnapshot();
      });

      it("Outer threshold is bigger than input rect", () => {
        expect(threshold.thresholds[CONTAINER.SK].bottom).toBeGreaterThan(
          CONTAINER.RECT.bottom
        );

        expect(threshold.thresholds[CONTAINER.SK].top).toBeLessThan(
          CONTAINER.RECT.top
        );

        expect(threshold.thresholds[CONTAINER.SK].left).toBeLessThan(
          CONTAINER.RECT.left
        );

        expect(threshold.thresholds[CONTAINER.SK].right).toBeGreaterThan(
          CONTAINER.RECT.right
        );
      });

      it("Calculate the threshold correctly", () => {
        expect(CONTAINER.RECT.bottom + pixels.y).toBe(
          threshold.thresholds[CONTAINER.SK].bottom
        );

        expect(CONTAINER.RECT.top - pixels.y).toBe(
          threshold.thresholds[CONTAINER.SK].top
        );

        expect(CONTAINER.RECT.right + pixels.x).toBe(
          threshold.thresholds[CONTAINER.SK].right
        );

        expect(CONTAINER.RECT.left - pixels.x).toBe(
          threshold.thresholds[CONTAINER.SK].left
        );
      });

      it("Define threshold for container", () => {
        expect(threshold.thresholds[CONTAINER.SK]).toStrictEqual(
          CONTAINER.THRESHOLD
        );

        expect(threshold.isOut[CONTAINER.SK]).toMatchObject(
          DEFAULT_IS_OUT_INDICATORS
        );
      });

      it("Throw when adding the same container key", () => {
        expect(() => {
          threshold.setContainerThreshold(
            CONTAINER.SK,
            INSERTION_LAYER.key,
            DRAGGED.DP,
            CONTAINER.RECT,
            UNIFIED_CONTAINER_DIMENSION
          );
        }).toThrow();
      });

      it("Define threshold for accumulated depth", () => {
        expect(threshold.thresholds[DRAGGED.DP]).toStrictEqual(DEPTH.THRESHOLD);

        expect(threshold.isOut[DRAGGED.DP]).toMatchObject(
          DEFAULT_IS_OUT_INDICATORS
        );
      });

      it("Define threshold for biggest container in the given depth", () => {
        expect(threshold.thresholds[INSERTION_LAYER.key]).toStrictEqual(
          INSERTION_LAYER.THRESHOLD
        );

        expect(threshold.isOut[INSERTION_LAYER.key]).toMatchObject(
          DEFAULT_IS_OUT_INDICATORS
        );
      });
    });
  });

  describe("Scroll Threshold", () => {
    let threshold: DFlexThreshold;

    const SCROLL = {
      ID_INNER: "scroll-id-inner",
      ID_OUTER: "scroll-id-outer",
      RECT: {
        top: 0,
        right: 885,
        bottom: 657,
        left: 0,
        height: 657,
        width: 885,
      },
      THRESHOLD_INNER: {
        top: 99,
        left: 133,
        bottom: 558,
        right: 752,
      },
      THRESHOLD_OUTER: {
        top: -99,
        left: -133,
        bottom: 756,
        right: 1018,
      },
    };

    const pixels = {
      x: 133,
      y: 99,
    };

    beforeAll(() => {
      const thresholdInput = {
        vertical: 15,
        horizontal: 15,
      };

      threshold = new DFlexThreshold(thresholdInput);
    });

    afterAll(() => {
      threshold.destroy();
    });

    describe("Scroll inner threshold", () => {
      it("Should be defined before setting the keys", () => {
        expect(threshold).toBeDefined();
        expect(threshold.isOut).toStrictEqual({});
        expect(threshold.thresholds).toStrictEqual({});
      });

      it("Add main inner threshold for scroll", () => {
        threshold.setMainThreshold(SCROLL.ID_INNER, SCROLL.RECT, true);

        expect(threshold).toMatchSnapshot();

        expect(threshold.isOut[SCROLL.ID_INNER]).toMatchObject(
          DEFAULT_IS_OUT_INDICATORS
        );

        expect(threshold.thresholds[SCROLL.ID_INNER]).toStrictEqual(
          SCROLL.THRESHOLD_INNER
        );
      });

      it("Inner threshold is less than input rect", () => {
        expect(threshold.thresholds[SCROLL.ID_INNER].bottom).toBeLessThan(
          SCROLL.RECT.top + SCROLL.RECT.height
        );

        expect(threshold.thresholds[SCROLL.ID_INNER].top).toBeGreaterThan(
          SCROLL.RECT.top
        );

        expect(threshold.thresholds[SCROLL.ID_INNER].left).toBeGreaterThan(
          SCROLL.RECT.left
        );

        expect(threshold.thresholds[SCROLL.ID_INNER].right).toBeLessThan(
          SCROLL.RECT.left + SCROLL.RECT.width
        );
      });

      it("Calculate the threshold correctly", () => {
        expect(SCROLL.RECT.top + SCROLL.RECT.height - pixels.y).toBe(
          threshold.thresholds[SCROLL.ID_INNER].bottom
        );

        expect(SCROLL.RECT.top + pixels.y).toBe(
          threshold.thresholds[SCROLL.ID_INNER].top
        );

        expect(SCROLL.RECT.left + SCROLL.RECT.width - pixels.x).toBe(
          threshold.thresholds[SCROLL.ID_INNER].right
        );

        expect(SCROLL.RECT.left + pixels.x).toBe(
          threshold.thresholds[SCROLL.ID_INNER].left
        );
      });
    });

    describe("Scroll outer threshold", () => {
      it("Add main outer threshold for scroll", () => {
        threshold.setMainThreshold(SCROLL.ID_OUTER, SCROLL.RECT, false);

        expect(threshold).toMatchSnapshot();

        expect(threshold.isOut[SCROLL.ID_OUTER]).toMatchObject(
          DEFAULT_IS_OUT_INDICATORS
        );

        expect(threshold.thresholds[SCROLL.ID_OUTER]).toStrictEqual(
          SCROLL.THRESHOLD_OUTER
        );
      });

      it("Outer threshold is less bigger than input rect", () => {
        expect(threshold.thresholds[SCROLL.ID_OUTER].bottom).toBeGreaterThan(
          SCROLL.RECT.top + SCROLL.RECT.height
        );

        expect(threshold.thresholds[SCROLL.ID_OUTER].top).toBeLessThan(
          SCROLL.RECT.top
        );

        expect(threshold.thresholds[SCROLL.ID_OUTER].left).toBeLessThan(
          SCROLL.RECT.left
        );

        expect(threshold.thresholds[SCROLL.ID_OUTER].right).toBeGreaterThan(
          SCROLL.RECT.left + SCROLL.RECT.width
        );
      });
    });
  });
});
