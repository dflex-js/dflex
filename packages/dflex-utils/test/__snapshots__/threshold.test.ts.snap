// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DFlexThreshold Dragged Threshold Container threshold Add threshold for container, it's inner by default 1`] = `
DFlexThreshold {
  "_percentages": Object {
    "horizontal": 60,
    "vertical": 60,
  },
  "_pixels": PointNum {
    "x": 102,
    "y": 30,
  },
  "isOut": Object {
    "0": FourDirectionsBool {
      "bottom": false,
      "left": false,
      "right": false,
      "top": false,
    },
    "0-container-sk": FourDirectionsBool {
      "bottom": false,
      "left": false,
      "right": false,
      "top": false,
    },
    "container-sk": FourDirectionsBool {
      "bottom": false,
      "left": false,
      "right": false,
      "top": false,
    },
    "dflex-dragged": FourDirectionsBool {
      "bottom": false,
      "left": false,
      "right": false,
      "top": false,
    },
  },
  "thresholds": Object {
    "0": Object {
      "bottom": 572,
      "left": -182,
      "right": 450,
      "top": 54,
    },
    "0-container-sk": Object {
      "bottom": 542,
      "left": -80,
      "right": 348,
      "top": 84,
    },
    "container-sk": Object {
      "bottom": 194,
      "left": -80,
      "right": 294,
      "top": 84,
    },
    "dflex-dragged": Object {
      "bottom": 194,
      "left": 348,
      "right": 722,
      "top": 84,
    },
  },
}
`;

exports[`DFlexThreshold Scroll Threshold Scroll inner threshold Add main inner threshold for scroll 1`] = `
DFlexThreshold {
  "_percentages": Object {
    "horizontal": 15,
    "vertical": 15,
  },
  "_pixels": PointNum {
    "x": 133,
    "y": 99,
  },
  "isOut": Object {
    "scroll-id-inner": FourDirectionsBool {
      "bottom": false,
      "left": false,
      "right": false,
      "top": false,
    },
  },
  "thresholds": Object {
    "scroll-id-inner": Object {
      "bottom": 558,
      "left": 133,
      "right": 752,
      "top": 99,
    },
  },
}
`;

exports[`DFlexThreshold Scroll Threshold Scroll outer threshold Add main outer threshold for scroll 1`] = `
DFlexThreshold {
  "_percentages": Object {
    "horizontal": 15,
    "vertical": 15,
  },
  "_pixels": PointNum {
    "x": 133,
    "y": 99,
  },
  "isOut": Object {
    "scroll-id-inner": FourDirectionsBool {
      "bottom": false,
      "left": false,
      "right": false,
      "top": false,
    },
    "scroll-id-outer": FourDirectionsBool {
      "bottom": false,
      "left": false,
      "right": false,
      "top": false,
    },
  },
  "thresholds": Object {
    "scroll-id-inner": Object {
      "bottom": 558,
      "left": 133,
      "right": 752,
      "top": 99,
    },
    "scroll-id-outer": Object {
      "bottom": 756,
      "left": -133,
      "right": 1018,
      "top": -99,
    },
  },
}
`;
